libs/entitlement/README.md

```diff
@@ -1,11 +1,192 @@
-### Entitlement library
+# Entitlements
 
-This module implements mechanisms to grant and check permissions under the _entitlements_ system.
+This module implements mechanisms to grant and check permissions under the _Entitlements_ system.
 
-The entitlements system provides an alternative to the legacy `SecurityManager` system, which is deprecated for removal.
-The `entitlement-agent` instruments sensitive class library methods with calls to this module, in order to enforce the controls.
+The entitlements system provides an alternative to the legacy Java Security Manager;
+Elasticsearch (ES) has previously made heavy use of the Java Security Manager to minimize the risk of security vulnerabilities impact. The Java Security Manager has been [deprecated for removal since Java 17](https://openjdk.org/jeps/411) (Sept 2021) and has been [removed in JDK 24](https://openjdk.org/jeps/486) (March 2025). Without an alternative, the removal of the Java Security Manager would have left Elasticsearch users more susceptible to future security vulnerabilities.
 
-This feature is currently under development, and it is completely disabled by default (the agent is not loaded). To enable it, run Elasticsearch with
+The goal of _entitlements_ is to protect certain sensitive operations on resources, and the JVM itself, from unexpected and unwanted access, e.g. to limit the scope of potential remote code execution (RCE) vulnerabilities.
+
+In practice, an entitlement allows code to call a well-defined set of corresponding JDK methods; without the entitlement code calling into those JDK methods is blocked and gets a `NotEntitledException`.
+
+## Structure
+
+All the code implementing Entitlements can be found under this directory. The `agent` module instruments sensitive JDK class library methods using a `InstrumentationService`. The current implementation of the instrumentation service uses ASM and is located under `asm-provider`.
```

Comment:
nit: the agent module doesn't do this, it's the entitlement lib. The agent module is just what loads/initializes the lib.